### PR TITLE
feat(plugin): add `subscription` and `publication` command groups

### PR DESCRIPTION
**Pull request by leonardoce**
Introduce support for new commands in the `cnpg` plugin for `kubectl` to enable native logical replication system integration with PostgreSQL. This includes the following commands:

- `publication create`
- `publication drop`
- `subscription create`
- `subscription drop`
- `subscription sync-sequences`

This marks our initial exploration into the native logical replication system of PostgreSQL. These commands empower users to imperatively perform online database import operations (from any PostgreSQL 10+ server) into a CloudNativePG-operated Postgres cluster, facilitating major online upgrades. All PostgreSQL logical replication limitations apply here.

Closes: #3994
Closes: #3978 
Closes: #4090

### CODE REVIEW

**Comment by leonardoce** [comment]
We still miss documentation and a few E2e tests, but this is lovely IMO

File: internal/cmd/plugin/logical/publication/create/cmd.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	var publicationName string
+	var externalClusterName string
+	var dryRun bool
+
+	publicationCreateCmd := &cobra.Command{
+		Use:   "create cluster_name",
+		Args:  cobra.ExactArgs(1),
+		Short: "create a logical replication publication",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			dbName := strings.TrimSpace(dbName)
```
Comment (gbartolini) [inline internal/cmd/plugin/logical/publication/create/cmd.go:43]: `dbName` should not be mandatory. We should allow users to pass the `--dbname` option, while defaulting to `app`.

File: internal/cmd/plugin/logical/subscription/syncsequences/cmd.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+		&externalClusterName,
+		"external-cluster",
+		"",
+		"The external cluster name",
+	)
+	syncSequencesCmd.Flags().StringVar(
+		&dbName,
+		"dbname",
+		"",
+		"The name of the application where to refresh the sequences. Defaults to the application database if available",
```
Comment (gbartolini) [inline internal/cmd/plugin/logical/subscription/syncsequences/cmd.go:118]:
```suggestion
		"The name of the database where the subscription is present and sequences need to be updated. Defaults to the application database, if available",
```

**Review by gbartolini — APPROVED** [review]

**Review by gbartolini — APPROVED** [review]

File: docs/src/kubectl-plugin.md [hunk 3]
```diff
@@ -1063,6 +1063,216 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+another one, even on an higher version of PostgreSQL. Before cutting over
+applications to the new database, sequences are updated.
+The `cnpg subscription sync-sequences` command connects to the database of
+source, pulls down all the sequences, and then updates the ones with the same
+identity (database schema and sequence name) locally.
+
+The command to execute is similar to the following:
+
+```sh
+kubectl-cnpg subscription sync-sequences \
```
Comment (fcanovai) [inline docs/src/kubectl-plugin.md:1223]:
```suggestion
kubectl cnpg subscription sync-sequences \
```

Comment (fcanovai) [reply docs/src/kubectl-plugin.md:1223]: Actually, the entire paragraph is repeated twice

**Review by fcanovai — CHANGES_REQUESTED** [review]

File: docs/src/kubectl-plugin.md (hunk 3 above)
Comment (gbartolini) [reply docs/src/kubectl-plugin.md:1223]: You are right. Let me fix it.

Comment (gbartolini) [reply docs/src/kubectl-plugin.md]: fixed

File: internal/cmd/plugin/logical/publication/create/cmd.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+			if len(publicationName) == 0 {
+				return fmt.Errorf("publication is a required option")
+			}
+
+			if allTables && (len(schemaNames) > 0 || len(tableExprs) > 0) {
+				return fmt.Errorf("cannot publicate all tables and selected schema/tables at the same time")
+			}
+			if !allTables && len(schemaNames) == 0 && len(tableExprs) == 0 {
+				return fmt.Errorf("no publication target selected")
+			}
```
Comment (fcanovai) [inline internal/cmd/plugin/logical/publication/create/cmd.go:68]: We could delegate this to cobra, using `MarkFlagRequired`, `MarkFlagsMutuallyExclusive` and `MarkFlagsOneRequired`. See https://github.com/spf13/cobra/blob/main/site/content/user_guide.md#required-flags

File: docs/src/kubectl-plugin.md [hunk 5]
```diff
@@ -1063,6 +1063,183 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+
+!!! Warning
+    When connecting to an external cluster, ensure that the specified user has
+    sufficient permissions to execute the `CREATE PUBLICATION` command.
+
+For additional information and detailed instructions, type the following
+command:
+
+```sh
+kubectl cnpg publication create help
```
Comment (jsilvela) [inline docs/src/kubectl-plugin.md:1120]:
```suggestion
kubectl cnpg publication create --help
```

File: docs/src/kubectl-plugin.md [hunk 6]
```diff
@@ -1063,6 +1063,183 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+
+This command configures a subscription directed towards the specified
+publication in the designated external cluster, as defined in the
+`<LOCAL_CLUSTER>`.
+
+For additional information and detailed instructions, type the following
+command:
+
+```sh
+kubectl cnpg subscription create help
```
Comment (jsilvela) [inline docs/src/kubectl-plugin.md:1183]:
```suggestion
kubectl cnpg subscription create --help
```

File: docs/src/kubectl-plugin.md [hunk 7]
```diff
@@ -1063,6 +1063,183 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+```sh
+kubectl cnpg subcription drop \
+  --subscription <SUBSCRIPTION_NAME> \
+  <LOCAL_CLUSTER> [options]
+```
+
+To access further details and precise instructions, use the following command:
+
+```sh
+kubectl cnpg subscription drop help
```
Comment (jsilvela) [inline docs/src/kubectl-plugin.md:1205]:
```suggestion
kubectl cnpg subscription drop --help
```

File: docs/src/kubectl-plugin.md [hunk 8]
```diff
@@ -1063,6 +1063,183 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+kubectl cnpg subscription sync-sequences \
+  --subscription <SUBSCRIPTION_NAME> \
+  <LOCAL_CLUSTER>
+```
+
+For comprehensive details and specific instructions, utilize the following
+command:
+
+```sh
+kubectl cnpg subscription sync-sequences help
```
Comment (jsilvela) [inline docs/src/kubectl-plugin.md:1235]:
```suggestion
kubectl cnpg subscription sync-sequences --help
```

File: internal/cmd/plugin/logical/publication/create/cmd.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+		&externalClusterName,
+		"external-cluster",
+		"",
+		"The cluster where to create the publication. Defaults to the local cluster",
+	)
+	publicationCreateCmd.Flags().BoolVar(
+		&dryRun,
+		"dry-run",
+		false,
+		"If specified, the publication is not created",
```
Comment (jsilvela) [inline internal/cmd/plugin/logical/publication/create/cmd.go:158]:
```suggestion
		"If specified, the publication commands are shown but not executed",
```

File: internal/cmd/plugin/logical/subscription/drop/drop.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		&dbName,
+		"dbname",
+		"",
+		"The database in which the command should drop the subscription",
+	)
+	subscriptionDropCmd.Flags().BoolVar(
+		&dryRun,
+		"dry-run",
+		false,
+		"If specified, the subscription is not deleted",
```
Comment (jsilvela) [inline internal/cmd/plugin/logical/subscription/drop/drop.go:79]:
```suggestion
		"If specified, the subscription deletion commands are shown but not executed",
```

**Review by jsilvela — COMMENTED** [review]
Still going over this but have left a few quick suggestions.
And, in my machine at least the `help` for kubectl will not work without `--`

File: docs/src/kubectl-plugin.md [hunk 11]
```diff
@@ -1063,6 +1063,183 @@ kubectl cnpg pgadmin4 --dry-run cluster-example | kubectl delete -f -
...
+step in this process involves updating sequences before transitioning
+applications to the new database (*cutover*).
+
+To address this limitation, the `cnpg subscription sync-sequences` command
+offers a solution. This command establishes a connection with the source
+database, retrieves all relevant sequences, and subsequently updates local
+sequences with matching identities (based on database schema and sequence
+name).
+
+You can the command as shown below:
```
Comment (jsilvela) [inline docs/src/kubectl-plugin.md:1223]:
```suggestion
You can use the command as shown below:
```

File: docs/src/samples/cluster-example-logical-dest.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      # created directly when connected to the target server.
+      # See cluster-example-logical-source.yaml for more information
+      # about this.
+      user: postgres
+      dbname: app
+    password:
+      name: cluster-example-superuser
+      key: password
+
+# To create the publication while connected to the target side, use:
```
Comment (jsilvela) [inline docs/src/samples/cluster-example-logical-dest.yaml:35]: "while connected" ... not sure what this means here

File: docs/src/samples/cluster-example-logical-source.yaml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        - ALTER TABLE numbers OWNER TO app;
+        - CREATE TABLE numbers_two (i SERIAL PRIMARY KEY, m INTEGER)
+        - INSERT INTO numbers_two (m) (SELECT generate_series(1,10000))
+        - ALTER TABLE numbers_two OWNER TO app;
+        - CREATE TABLE numbers_three (i SERIAL PRIMARY KEY, m INTEGER)
+        - INSERT INTO numbers_three (m) (SELECT generate_series(1,10000))
+        - ALTER TABLE numbers_three OWNER TO app;
+
+  # This is needed to enable users to create the publication while
+  # they are connected from the destination. Like in:
```
Comment (jsilvela) [inline docs/src/samples/cluster-example-logical-source.yaml:27]: again, "while they are connected".
The commands below can be executed from the plain shell, you don't need to be connected before.
I think what is meant is something like
"to create the publication by issuing commands to be executed on the destination cluster"

File: docs/src/samples/cluster-example-logical-dest.yaml (hunk 12 above)
Comment (jsilvela) [reply docs/src/samples/cluster-example-logical-dest.yaml:35]: The commands are abstracting us from where they get executed.

File: internal/cmd/plugin/logical/publication/create/cmd.go [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+		&tableExprs,
+		"table",
+		nil,
+		"Create the publication for the selected table expression",
+	)
+	publicationCreateCmd.Flags().StringVar(
+		&externalClusterName,
+		"external-cluster",
+		"",
+		"The cluster where to create the publication. Defaults to the local cluster",
```
Comment (jsilvela) [inline internal/cmd/plugin/logical/publication/create/cmd.go:152]:
```suggestion
		"The cluster in which to create the publication. Defaults to the local cluster",
```

File: internal/cmd/plugin/logical/subscription/drop/drop.go (hunk 10 above)
Comment (leonardoce) [reply internal/cmd/plugin/logical/subscription/drop/drop.go]: Done in a separate commit.

File: docs/src/samples/cluster-example-logical-dest.yaml (hunk 12 above)
Comment (leonardoce) [reply docs/src/samples/cluster-example-logical-dest.yaml]: I removed the cnpg invocation from the examples. They're too far away from the code to be really kept updated.

**Review by jsilvela — APPROVED** [review]

File: internal/cmd/plugin/logical/subscription/drop/cmd.go [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+	)
+	_ = subscriptionDropCmd.MarkFlagRequired("subscription")
+
+	subscriptionDropCmd.Flags().StringVar(
+		&dbName,
+		"dbname",
+		"",
+		"The database in which the command should drop the subscription (required)",
+	)
+	_ = subscriptionDropCmd.MarkFlagRequired("dbname")
```
Comment (fcanovai) [inline internal/cmd/plugin/logical/subscription/drop/cmd.go:74]: This is not the same behaviour of the other commands, which default to the application database if available.

**Review by fcanovai — APPROVED** [review]

**Pull request merged**